benches/throughput.rs

```diff
@@ -0,0 +1,144 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::env;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::sync::Mutex;
+
+use criterion::{
+    criterion_group, criterion_main, BenchmarkId, Criterion, SamplingMode, Throughput,
+};
+use pprof::criterion::{Output, PProfProfiler};
+use tokio::net::TcpStream;
+use tokio::runtime::Runtime;
+
+use tracing::info;
+
+use ztunnel::test_helpers::app::TestApp;
+use ztunnel::test_helpers::{echo, helpers, tcp};
+use ztunnel::{app, identity, test_helpers};
+
+const KB: usize = 1024;
+const MB: usize = 1024 * KB;
+
+struct TestEnv {
+    direct: TcpStream,
+    tcp: TcpStream,
+    hbone: TcpStream,
+}
+
+/// initialize_environment sets up a benchmarking environment. This works around issues in async setup with criterion.
+/// Since tests are only sending data on existing connections, we setup a connection for each test type in the setup phase.
+/// Tests consume the
+fn initialize_environment() -> (Arc<Mutex<TestEnv>>, Runtime) {
+    if env::var("RUST_LOG").is_err() {
+        env::set_var("RUST_LOG", "error")
+    }
+    helpers::initialize_telemetry();
+    let rt = tokio::runtime::Builder::new_multi_thread()
+        .enable_all()
+        .build()
+        .unwrap();
+    // Global setup: spin up an echo server and ztunnel instance
+    let (env, _) = rt.block_on(async move {
+        // let mut env = async_global_setup(|| async move {
+        let cert_manager = identity::mock::MockCaClient::new(Duration::from_secs(10));
+        let app = app::build_with_cert(test_helpers::test_config(), cert_manager)
+            .await
+            .unwrap();
+
+        let ta = TestApp {
+            admin_address: app.admin_address,
+            proxy_addresses: app.proxy_addresses,
+        };
+        ta.ready().await;
+        let echo = echo::TestServer::new().await;
+        let echo_addr = helpers::with_ip(echo.address(), "127.0.0.1".parse().unwrap());
+        let t = tokio::spawn(async move {
+            let _ = tokio::join!(app.spawn(), echo.run());
+        });
+        let mut hbone = ta
+            .socks5_connect(helpers::with_ip(echo_addr, "127.0.0.1".parse().unwrap()))
+            .await;
+        let mut tcp = ta
+            .socks5_connect(helpers::with_ip(echo_addr, "127.0.0.2".parse().unwrap()))
+            .await;
+        let mut direct = TcpStream::connect(echo_addr).await.unwrap();
+        direct.set_nodelay(true).unwrap();
+        info!("setup complete");
+
+        // warmup: send 1 byte so we ensure we have the full connection setup.
+        tcp::run_latency(&mut hbone, 1).await.unwrap();
```

Comment:
Should we consider lazily doing this? Probably not needed, but it feels weird to have the unused ones idling.

Reply:
Getting setup properly for our needs with criterion seems to be really hard. I guess making it Lazy would actually probably do the trick.. we could do that.

Or have 1 full setup per type (direct, tcp, hbone) and only establish one connection.